pkg/api/v1/publish-api-types.go

```diff
@@ -0,0 +1,248 @@
+// and the 'code' field will contain one of the constants defined in this file.
+// The intent is that code can be used to show a localized error message on the
+// device.
+//
+// The Padding field may be populated with random data on both success and
+// error responses.
+type PublishKeysResponseServer struct {
+	RevisionToken     string `json:"revisionToken"`
+	InsertedExposures int    `json:"insertedExposures"`
+	ErrorMessage      string `json:"error"`
+	Code              string `json:"code"`
+	Padding           string `json:"padding"`
+}
+
+// ServerExposureKey is the 16 byte key, the start time of the key and the duration of the key. A duration of 0 means 24 hours.
+// ALL fields are REQUIRED and must meet the constraints below.
+//
+// Key must be the base64 (RFC 4648) encoded 16 byte exposure key from the device.
+// - Base64 encoding should include padding, as per RFC 4648
+// - if the key is not exactly 16 bytes in length, the request will be failed
+// - that is, the whole batch will fail.
+// IntervalNumber must be "reasonable" as in the system won't accept keys that
+//   are scheduled to start in the future or that are too far in the past, which
+//   is configurable per installation.
+//
+// IntervalCount must >= `minIntervalCount` and <= `maxIntervalCount`; 1 - 144 inclusive.
+//
+// TransmissionRisk must be >= 0 and <= 8.
+//   Transmission risk is optional, but should still be populated for compatibility
+//   with older clients. If it is omitted, and there is a valid report type,
+//   then transmissionRisk will be set to 0.
+//   IF there is a report type from the verification certificate AND tranismission risk
+//    is not set, then a report type of
+//     CONFIRMED will lead to TR 2
+//     LIKELY will lead to TR 4
+//     NEGATIVE will lead to TR 6
+type ServerExposureKey struct {
+	Key              string `json:"key"`
+	IntervalNumber   int32  `json:"rollingStartNumber"`
+	IntervalCount    int32  `json:"rollingPeriod"`
+	TransmissionRisk int    `json:"transmissionRisk,omitempty"` // Optional
+}
+
+//ReportType means type of the keys report - is it self-report, confirmed diagnose, ...?
+type ReportType string
+
+// Possible values of ReportType.
+const (
+	Unknown                    = "Unknown"
+	ConfirmedTest              = "ConfirmedTest"
+	ConfirmedClinicalDiagnosis = "ConfirmedClinicalDiagnosis"
+	SelfReport                 = "SelfReport"
+	Recursive                  = "Recursive"
+	Revoked                    = "Revoked"
+)
+
+//UnmarshalJSON checks ReportType for having a valid or empty value.
+func (lt *ReportType) UnmarshalJSON(b []byte) error {
+	var s string
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+
+	reportType := ReportType(s)
+	switch reportType {
+	case Unknown, ConfirmedTest, ConfirmedClinicalDiagnosis, SelfReport, Recursive, Revoked:
+		*lt = reportType
+		return nil
+	}
+	return errors.New("Invalid leave type")
+}
```

Comment:
isn't this more like some internals/utils.go rather than public api?:)

Reply:
I don't know. I was that much pissed Go lack so basic feature I didn't really think about that. But you're probably right...